### PR TITLE
Sabre with measurements

### PR DESCRIPTION
**Pull request by Simone-Bordoni**
Checklist:
- [x] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

### CODE REVIEW

**Comment by Simone-Bordoni** [comment]
To make the Sabre transpiler work with qubits i have already modified the block decomposition to accept measurements.
Now measurements acting on more than one qubits are decomposed into single qubit measurements so that they do not interfere with the block decomposition. For now i managed to match the .result of new single qubit measurements that was having problems with the block.on_qubits method. another way to solve the problem would be to overload the .on_qubits method for measurement gates in order to return a gate pointing at the same result object. Let me know if you prefer this implementation or the one implemented now. in order to match the results for multi qubit measurement gates it will be a bit more difficult.
@stavros11 @BrunoLiegiBastonLiegi

**Comment by Simone-Bordoni** [comment]
The Sabre transpiler with this PR will be able to corrrectly work with measurement gates. In particular final measurements can be given on multi qubits and registers are correctly matched. Intermediate measurements works correctly if acting on single qubits. When using multi qubit intermediate measurements they will be decomposed into single qubit measurements thus changing their register. However, as discussed with @stavros, for now intermediate measurements are not suported on the hardware, so this issue can be addressed in the future.
As reported in #1094 there are problems with the measurement registers when both intermediate and final measurements are present in the circuit, when this issue will be solved alse sabre should be able to deal with circuits with both kind of measurements.

File: src/qibo/transpiler/router.py [hunk 1]
```diff
@@ -371,9 +371,13 @@ def execute_block(self, block: Block):
         self._routed_blocks.add_block(block.on_qubits(self.get_physical_qubits(block)))
         self.circuit_blocks.remove_block(block)
 
-    def routed_circuit(self):
-        """Returns circuit of the routed circuit."""
-        return self._routed_blocks.circuit()
+    def routed_circuit(self, circuit_kwargs=None):
+        """Returns circuit of the routed circuit.
```
Comment (BrunoLiegiBastonLiegi) [inline src/qibo/transpiler/router.py:375]:
```suggestion
        """Returns the routed circuit.
```

File: src/qibo/transpiler/router.py [hunk 2]
```diff
@@ -485,22 +496,63 @@ def added_swaps(self):
...
+        new_circuit = Circuit(circuit.nqubits)
+        for gate in circuit.queue:
+            new_circuit.add(gate)
+        return new_circuit
+
+    def _detach_final_measurements(self, circuit: Circuit):
+        """Detach measurement gates at the end of the circuit for separate handling."""
+        final_measurements = []
+        reversed_queue = circuit.queue[::-1]
+        for gate in reversed_queue:
```
Comment (BrunoLiegiBastonLiegi) [inline src/qibo/transpiler/router.py:530]:
```suggestion
        for gate in circuit.queue[::-1]:
```

File: src/qibo/transpiler/router.py [hunk 3]
```diff
@@ -485,22 +496,63 @@ def added_swaps(self):
...
+            if isinstance(gate, gates.M):
+                final_measurements.append(gate)
+                circuit.queue.remove(gate)
+            else:
+                break
+        print(final_measurements)
```
Comment (BrunoLiegiBastonLiegi) [inline src/qibo/transpiler/router.py]:
```suggestion
```

File: src/qibo/transpiler/router.py [hunk 4]
```diff
@@ -485,22 +496,63 @@ def added_swaps(self):
...
+        if len(final_measurements) == 0:
```
Comment (BrunoLiegiBastonLiegi) [inline src/qibo/transpiler/router.py:536]:
```suggestion
        if not final_measurements:
```

File: src/qibo/transpiler/router.py [hunk 5]
```diff
@@ -485,22 +496,63 @@ def added_swaps(self):
...
+            return None
+        return final_measurements[::-1]
+
+    def _append_final_measurements(self, routed_circuit: Circuit):
+        """Append the final measurment gates on the correct qubits conserving the measurement register."""
+        for measurement in self._final_measurements:
+            print(measurement)
```
Comment (BrunoLiegiBastonLiegi) [inline src/qibo/transpiler/router.py]:
```suggestion
```

File: tests/test_transpiler_router.py [hunk 6]
```diff
@@ -246,7 +246,7 @@ def test_circuit_map():
     circuit_map.execute_block(block_list.search_by_index(1))
     circuit_map.execute_block(block_list.search_by_index(2))
     circuit_map.execute_block(block_list.search_by_index(3))
-    routed_circuit = circuit_map.routed_circuit()
+    routed_circuit = circuit_map.routed_circuit(circuit_kwargs=None)
```
Comment (BrunoLiegiBastonLiegi) [inline tests/test_transpiler_router.py:249]: `circuit_kwargs` should be `None` by default, right?

File: tests/test_transpiler_router.py [hunk 7]
```diff
@@ -331,3 +331,36 @@ def test_sabre_memory_map():
...
+def test_sabre_intermediate_measurements():
+    measurement = gates.M(0)
+    circ = Circuit(3, density_matrix=True)
+    circ.add(gates.H(0))
+    circ.add(measurement)
+    circ.add(gates.CNOT(0, 1))
+    connectivity = nx.Graph()
+    connectivity.add_nodes_from([0, 1, 2])
+    connectivity.add_edges_from([(0, 1), (1, 2)])
+    router = Sabre(connectivity=connectivity)
```
Comment (BrunoLiegiBastonLiegi) [inline tests/test_transpiler_router.py:345]: But here Sabre is doing nothing right? The CNOT can be executed without any SWAP. I would test the scenery where the measurement is actually moved by the routing.

**Review by BrunoLiegiBastonLiegi — COMMENTED** [review]
Thanks @Simone-Bordoni, just a handful of comments.

**Review by BrunoLiegiBastonLiegi — COMMENTED** [review]
In the `test_sabre_intermediate_measurements` the transpiler is doing nothing I believe, as the `CNOT(0,1)` involves qubits that are connected already. I tried to change this to `CNOT(0,2)` to trigger the `SWAP` insertion and it fails for me. I would change that test by adding the measurement to qubit 1 `M(1)` and applying `CNOT(0,2)`. This way you are sure that the measurement is moved and sabre is doing something.
The `test_sabre_final_measurements` I would remove, instead, and add the final measurements testing to `test_sabre_random_circuits`.
```
@pytest.mark.parametrize("gates", [10, 40])
@pytest.mark.parametrize("look", [0, 5])
@pytest.mark.parametrize("decay", [0.5, 1.0])
@pytest.mark.parametrize("placer", [Trivial, Random])
@pytest.mark.parametrize("connectivity", [star_connectivity(), grid_connectivity()])
def test_sabre_random_circuits(gates, look, decay, placer, connectivity):
    placer = placer(connectivity=connectivity)
    layout_circ = Circuit(5)
    initial_layout = placer(layout_circ)
    router = Sabre(connectivity=connectivity, lookahead=look, decay_lookahead=decay)
    circuit = generate_random_circuit(nqubits=5, ngates=gates)
    measurement = gates.M(np.random.choice(range(5), size=3, replace=False))
    circuit.add(measurement)
    transpiled_circuit, final_qubit_map = router(circuit, initial_layout)
    assert router.added_swaps >= 0
    assert_connectivity(connectivity, transpiled_circuit)
    assert_placement(transpiled_circuit, final_qubit_map)
    assert gates + router.added_swaps == transpiled_circuit.ngates
    assert_circuit_equivalence(
        original_circuit=circuit,
        transpiled_circuit=transpiled_circuit,
        final_map=final_qubit_map,
        initial_map=initial_layout,
    )
    circuit_result = routed_circ.execute(nshots=100)
    assert circuit_result.frequencies() == measurement.result.frequencies()
    assert routed_circ.queue[-1].result is measurement.result
```

**Comment by BrunoLiegiBastonLiegi** [comment]
In `test_intermediate_measurements` I tried moving the measurement to qubit 1, i.e. the one that gets swapped by the router, and it's moved at the end of the queue after routing.
```
# circ.queue
[<qibo.gates.gates.H object at 0x7f0f2beffb90>, <qibo.gates.measurements.M object at 0x7f0f2be5dc50>, <qibo.gates.gates.CNOT object at 0x7f0f2beffbd0>]

# routed_circ.queue
[<qibo.gates.gates.SWAP object at 0x7f0f285287d0>, <qibo.gates.gates.H object at 0x7f0f2852b110>, <qibo.gates.gates.CNOT object at 0x7f0f2852b2d0>, <qibo.gates.measurements.M object at 0x7f0f2bef3d90>]
```

**Comment by Simone-Bordoni** [comment]
> In `test_intermediate_measurements` I tried moving the measurement to qubit 1, i.e. the one that gets swapped by the router, and it's moved at the end of the queue after routing.
> 
> ```
> # circ.queue
> [<qibo.gates.gates.H object at 0x7f0f2beffb90>, <qibo.gates.measurements.M object at 0x7f0f2be5dc50>, <qibo.gates.gates.CNOT object at 0x7f0f2beffbd0>]
> 
> # routed_circ.queue
> [<qibo.gates.gates.SWAP object at 0x7f0f285287d0>, <qibo.gates.gates.H object at 0x7f0f2852b110>, <qibo.gates.gates.CNOT object at 0x7f0f2852b2d0>, <qibo.gates.measurements.M object at 0x7f0f2bef3d90>]
> ```

Thme measurement gate commutes with the CNOT (different qubits) so the order is irrelevant.

**Review by BrunoLiegiBastonLiegi — APPROVED** [review]

**Review by stavros11 — APPROVED** [review]

**Pull request merged**